### PR TITLE
Handle long tagged NSStrings

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

File: test/stdlib/NSSlowTaggedLocalizedString.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// REQUIRES: objc_interop
+
+import NSSlowTaggedLocalizedString
+import Swift
+
+import StdlibUnittest
+
+let longTaggedTests = TestSuite("NonContiguousTaggedStrings")
+
+longTaggedTests.test("EqualLongTagged") {
```
Comment (Catfish-Man) [inline test/stdlib/NSSlowTaggedLocalizedString.swift:17]: Aside from being absolutely horrific, this is failing for me locally with this error: "error: type of expression is ambiguous without more context"

I'm completely at a loss for why that would be happening, suggestions welcome.

**Comment by Catfish-Man** [comment]
@swift-ci please test

File: test/stdlib/NSSlowTaggedLocalizedString.swift (hunk 1 above)
Comment (LucianoPAlmeida) [reply test/stdlib/NSSlowTaggedLocalizedString.swift:17]: Not sure if is possible, but use `-debug-constraints` or `debug-constraints-on-line=${line}` and dump the type checker debug output into a file we would be able to get a hint into what type checker is doing and where it is getting lost? Hope that helps =]

File: stdlib/public/core/SmallString.swift [hunk 2]
```diff
@@ -353,13 +353,22 @@ extension _SmallString {
   //
   @_effects(readonly) // @opaque
   @usableFromInline // testable
-  internal init(taggedCocoa cocoa: AnyObject) {
+  internal init?(taggedCocoa cocoa: AnyObject) {
```
Comment (glessard) [inline stdlib/public/core/SmallString.swift:356]: ABI change! We probably need to add this initializer as `@_aEIC`, and mark the old one as unavailable.

Comment (Catfish-Man) [reply stdlib/public/core/SmallString.swift:356]: Argh, you’re right, it’s usableFromInline I missed that. Thanks!

File: stdlib/public/core/SmallString.swift [hunk 3]
```diff
@@ -353,13 +353,22 @@ extension _SmallString {
...
+      if let len = _bridgeTagged(cocoa, intoUTF8: $0) {
+        return len
+      }
+      success = false
+      return 0
+    }
+    if !success {
+      return nil
     }
     self._invariantCheck()
```
Comment (kperryua) [inline stdlib/public/core/SmallString.swift:373]: Does this `_invariantCheck()` need to get called even when we return `nil`?

Comment (glessard) [reply stdlib/public/core/SmallString.swift:373]: It only needs to get called if a `_SmallString` is actually built. None is when returning `nil`.

File: test/stdlib/NSSlowTaggedLocalizedString.swift (hunk 1 above)
Comment (Catfish-Man) [reply test/stdlib/NSSlowTaggedLocalizedString.swift]: This helped me figure it out! Thanks again :)

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

File: test/stdlib/Inputs/NSSlowTaggedLocalizedString/NSSlowTaggedLocalizedString.m [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+- (uint64_t)length {
+  return strlen(contents);
+}
+
+- (id)copyWithZone:(id)unused {
+  return self;
+}
+
+- (uint16_t)characterAtIndex:(NSUInteger)index {
+  abort();
```
Comment (kperryua) [inline test/stdlib/Inputs/NSSlowTaggedLocalizedString/NSSlowTaggedLocalizedString.m:81]: This was probably not intended.

File: test/stdlib/Inputs/NSSlowTaggedLocalizedString/NSSlowTaggedLocalizedString.m [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+- (uint64_t)_cfTypeID {
+  return 7; //CFString
+}
+
+- (id) description {
+  return self;
+}
+
+- (BOOL)getBytes:(void *)buffer maxLength:(uint64_t)max usedLength:(uint64_t *)used encoding:(uint64_t)encoding options:(uint64_t)options range:(NSRange)range remainingRange:(NSRange *)leftover {
+  strncpy(buffer, contents, max);
```
Comment (kperryua) [inline test/stdlib/Inputs/NSSlowTaggedLocalizedString/NSSlowTaggedLocalizedString.m:105]: Probably want to assert the encoding as either UTF8 or ASCII?

**Comment by Catfish-Man** [comment]
Updated status: new test passes locally with DYLD_LIBRARY_PATH when run manually, but fails when run normally and fails on the bot. Karoy and I are going to take a look tomorrow.

File: stdlib/public/core/SmallString.swift (hunk 2 above)
Comment (Catfish-Man) [reply stdlib/public/core/SmallString.swift:356]: We discussed this and I'm pretty sure we can get away with it: nothing actually inlines this

**Review by kperryua — APPROVED** [review]

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Review by lorentey — APPROVED** [review]

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Review by glessard — APPROVED** [review]

**Pull request merged**